### PR TITLE
Idempotence all grant

### PR DESCRIPTION
**Pull request by azeliashchonak**
##### SUMMARY

In MySQL 8.0 SHOW GRANTS no longer displays ALL PRIVILEGES in its global-privileges output. Instead, SHOW GRANTS explicitly lists each granted global privilege. The issue is that the mysql_user module is using a simple comparison between new and current privileges and if we define ".:ALL,GRANT" it will return changed, because "ALL" != list_of_privileges, so idempotence check will be failed.

##### ISSUE TYPE

Bugfix Pull Request

##### COMPONENT NAME

mysql_user

##### ADDITIONAL INFORMATION

Instead of https://github.com/ansible/ansible/pull/52469

### CODE REVIEW

File: lib/ansible/modules/database/mysql/mysql_user.py [hunk 1]
```diff
@@ -102,10 +102,8 @@
      the new root credentials. Subsequent runs of the playbook will then succeed by reading the new credentials from
      the file."
    - Currently, there is only support for the `mysql_native_password` encrypted password hash module.
-
 author:
 - Jonathan Mainguy (@Jmainguy)
-- Benjamin Malynovytch (@bmalynovytch)
```
Comment (gundalow) [inline lib/ansible/modules/database/mysql/mysql_user.py]: Why has this line been removed?

File: lib/ansible/modules/database/mysql/mysql_user.py [hunk 2]
```diff
@@ -605,9 +583,9 @@ def main():
             connect_timeout=dict(type='int', default=30),
             config_file=dict(type='path', default='~/.my.cnf'),
             sql_log_bin=dict(type='bool', default=True),
-            client_cert=dict(type='path', aliases=['ssl_cert']),
-            client_key=dict(type='path', aliases=['ssl_key']),
-            ca_cert=dict(type='path', aliases=['ssl_ca']),
+            ssl_cert=dict(type='path'),
```
Comment (gundalow) [inline lib/ansible/modules/database/mysql/mysql_user.py]: Not sure why these aliases have been swapped around?
This change will break people using `client_cert:` in their playbooks

**Review by gundalow — CHANGES_REQUESTED** [review]
Is it possible your branch wasn't up to date before creating this PR, I see lots of unrelated changed in here

See https://docs.ansible.com/ansible/devel/dev_guide/developing_rebasing.html for guidance on how to rebase your branch

File: lib/ansible/modules/database/mysql/mysql_user.py (hunk 1 above)
Comment (azeliashchonak) [reply lib/ansible/modules/database/mysql/mysql_user.py]: copy-paste old file) I'll fix it.

File: lib/ansible/modules/database/mysql/mysql_user.py (hunk 2 above)
Comment (azeliashchonak) [reply lib/ansible/modules/database/mysql/mysql_user.py]: fixed

**Comment by tgadiev** [comment]
@bmalynovytch can you take a look please?

File: lib/ansible/modules/database/mysql/mysql_user.py [hunk 3]
```diff
@@ -115,87 +115,74 @@
...
     host: localhost
     state: absent
-
 - name: Removes all anonymous user accounts
   mysql_user:
     name: ''
     host_all: yes
     state: absent
-
 - name: Create database user with name 'bob' and password '12345' with all database privileges
```
Comment (bmalynovytch) [inline lib/ansible/modules/database/mysql/mysql_user.py]: @azeliashchonak 
Many whitespace changes there, probably introduced by automatic formatting.
Could you revert those changes, unless they are necessary / related to the object of the current PR ?

File: lib/ansible/modules/database/mysql/mysql_user.py [hunk 4]
```diff
@@ -479,7 +471,8 @@ def pick(x):
             return x
 
     for grant in grants:
-        res = re.match("""GRANT (.+) ON (.+) TO (['`"]).*\\3@(['`"]).*\\4( IDENTIFIED BY PASSWORD (['`"]).+\5)? ?(.*)""", grant[0])
+        res = re.match(
+            """GRANT (.+) ON (.+) TO (['`"]).*\\3@(['`"]).*\\4( IDENTIFIED BY PASSWORD (['`"]).+\5)? ?(.*)""", grant[0])
```
Comment (bmalynovytch) [inline lib/ansible/modules/database/mysql/mysql_user.py]: There seem to be a regression there 🤔 : `\5)` should be `\6)` as far as I remember

File: lib/ansible/modules/database/mysql/mysql_user.py [hunk 5]
```diff
@@ -493,14 +486,28 @@ def pick(x):
...
+    privileges = []
+    cursor.execute('SELECT DISTINCT PRIVILEGE_TYPE FROM INFORMATION_SCHEMA.USER_PRIVILEGES WHERE IS_GRANTABLE="yes"')
+    grants = cursor.fetchall()
+    for grant in grants:
+        privileges.append(grant[0])
+    return privileges
+
+
+def get_db_version(cursor):
+    cursor.execute("Select @@version")
```
Comment (bmalynovytch) [inline lib/ansible/modules/database/mysql/mysql_user.py]:
```suggestion
    cursor.execute("SELECT @@version")
```

**Review by bmalynovytch — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/database/mysql/mysql_user.py (hunk 4 above)
Comment (bmalynovytch) [reply lib/ansible/modules/database/mysql/mysql_user.py]: I confirm: current version in devel reverted this change, probably when I handled conflicts on my own changes.

Comment (bmalynovytch) [reply lib/ansible/modules/database/mysql/mysql_user.py]: Fix in review https://github.com/ansible/ansible/pull/55292

File: lib/ansible/modules/database/mysql/mysql_user.py (hunk 3 above)
Comment (azeliashchonak) [reply lib/ansible/modules/database/mysql/mysql_user.py]: Why should we leave empty rows in comments?

File: lib/ansible/modules/database/mysql/mysql_user.py (hunk 4 above)
Comment (azeliashchonak) [reply lib/ansible/modules/database/mysql/mysql_user.py]: fixed

File: lib/ansible/modules/database/mysql/mysql_user.py (hunk 5 above)
Comment (azeliashchonak) [reply lib/ansible/modules/database/mysql/mysql_user.py]: fixed

File: lib/ansible/modules/database/mysql/mysql_user.py (hunk 3 above)
Comment (bmalynovytch) [reply lib/ansible/modules/database/mysql/mysql_user.py]: Oh ! We both misread / misunderstood theses changes: they are not whitespace changes, neither empty rows.
Removing these lines completely changes the structure, removing one level of list (list of lists becomes simple list).
I don't know if this extra level is necessary or not, we need to check how documentation should be provided.

@felixfontein @gundalow Do you have any input about this ?

Comment (felixfontein) [reply lib/ansible/modules/database/mysql/mysql_user.py]: @azeliashchonak this is not a comment, but extracted and put into the documentation of the module. See here: https://docs.ansible.com/ansible/devel/modules/mysql_user_module.html#examples

In general, examples are better readable if different examples are separated with empty lines. So I would definitely keep them.

Comment (azeliashchonak) [reply lib/ansible/modules/database/mysql/mysql_user.py]: I've got it, Thanks. Fixed.

**Comment by tgadiev** [comment]
@bmalynovytch can you take a look, pls?

**Comment by Andersson007** [comment]
@azeliashchonak , please solve conflicts or close this PR

needs_info

**Comment by bmalynovytch** [comment]
Branch primarily needs a rebase (which will require to solve conflicts)

**Pull request closed without merging**